### PR TITLE
Wizard changes are not forgotten when pressing cancel

### PR DESCRIPTION
**Pull request by despo**
Includes #5464 which we will be merging alongside this, so please only review last 3 commits here.

## Context

Clear wizard cache on entry.

This offers two solutions. The optimal solution checks that the referer is not part of the wizard flow routes. However, as that can't be applied on all wizards due to them not using resourceful routes and `Rails.application.routes.recognize_path(referer_path)` not being able to deconstruct the path, a backup solution is oferred where the wizard entrypoints are specified instead. This is not optimal but it is the most common use case. 

## Changes proposed in this pull request

Introduce a new module that can be mixed in the wizard controllers and with minimal configuration can enable clearing the wizard without having to do extra work.

## Guidance to review
One solution is used in the interview process, so 

1. begin setting up a new interview and after you get to the check page cancel and exit the process. Restart creating a new interview. The form should not be prefilled.

2. Click Make a decision and choose Reject application. Fill in reasons for rejection and get to the next step. Cancel the process. Attempt to restart the process. There should be no prefilled information.

## Link to Trello card
https://trello.com/c/N1DAdjHN/3821-wizard-changes-are-not-forgotten-when-pressing-cancel

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/controllers/concerns/clear_wizard_cache.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+module ClearWizardCache
+  include ActiveSupport::Concern
+
+private
+
+  def clear_wizard_if_new_entry(wizard, entrypoint_check: false)
+    return if (entrypoint_check && !entrypoint_path?) ||
+              (!entrypoint_check && entrypoint_in_wizard_flow?(wizard_flow_controllers, wizard_controller_excluded_paths))
```
Comment (benswannack) [inline app/controllers/concerns/clear_wizard_cache.rb:8]: What's the reason we have two options here?

I think the entrypoint_path one makes sense, where we specify which paths are the entrypoints, and clear the wizard when we come from them. I don't really understand the other one though

File: app/controllers/provider_interface/condition_statuses_controller.rb [hunk 2]
```diff
@@ -74,5 +68,13 @@ def success_message
         'Status of conditions updated'
       end
     end
+
+    def wizard_controller_excluded_paths
+      []
+    end
+
+    def wizard_flow_controllers
+      ['provider_interface/condition-statuses', 'provider_interface/condition_statuses/checks'].freeze
```
Comment (benswannack) [inline app/controllers/provider_interface/condition_statuses_controller.rb:77]: Should this be?
```suggestion
      ['provider_interface/condition_statuses', 'provider_interface/condition_statuses/checks'].freeze
```

File: app/controllers/concerns/clear_wizard_cache.rb (hunk 1 above)
Comment (benswannack) [reply app/controllers/concerns/clear_wizard_cache.rb:8]: Ok, discussed in person.

The way we wanted to do this is to specify a controller (or list of controllers) that make up the wizard flow, and reset the wizard if the referrer is not from the specified controllers.

The first caveat is that not all actions in wizard controllers are part of the flow e.g. `interviews#index` is not in the interview edit or create flow. Therefore we add  a list of paths which are from the specified controllers but not in the wizard flows.

The second caveat is that if routes are not set up in a particular way using `resources`, then this controller detection method does not work. So for more complicated controller setups, like reasons for rejection, we just specify which routes are the entrypoints for the flow, and clear the wizard if the referrer is one of those.

File: app/controllers/provider_interface/condition_statuses_controller.rb (hunk 2 above)
Comment (despo) [reply app/controllers/provider_interface/condition_statuses_controller.rb]: Good catch. It was causing the wizard to clear on refresh which is now fixed.

File: spec/system/wizard_cache_cleared_spec.rb [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    expect(page.find_field('Year').value).to be_nil
+    expect(page.find_field('Time').value).to be_nil
+    expect(page.find_field('Address or online meeting details').value).to be_empty
+  end
+
+  def then_i_should_be_on_the_decision_page
+    expect(page).to have_current_path(new_provider_interface_application_choice_decision_path(application_choice))
+  end
+
+  def and_i_seleet_why_i_am_rejecting_the_application
```
Comment (cwrw) [inline spec/system/wizard_cache_cleared_spec.rb:124]:
```suggestion
  def and_i_select_why_i_am_rejecting_the_application
```

**Review by cwrw — COMMENTED** [review]
looks great! will do another round of testing this morning when the last change goes in

**Review by cwrw — APPROVED** [review]
looks great! thanks for this @despo, this was def tricky 👍

**Pull request merged**